### PR TITLE
[ new ] Chapter on details of QTT

### PR DESCRIPTION
**Pull request by kiana-S**
A while ago (about a month ago at this point, oops) I offered to contribute a section in this tutorial explaining Quantitative Type Theory in depth. I've finally gotten around to typing up a draft of it, so here it is!

A few things to note:

* I didn't see any particularly obvious place to put this section in the table of contents, so I just left it unlinked.
* The new file `src/Tutorial/QTT.md` will fail to typecheck as a literate Idris file, so I didn't put it in the ipkg module list.

Edits to address any of these issues would be great!

### CODE REVIEW

**Comment by stefan-hoeck** [comment]
Thanks a lot for adding this! I will have to reread the whole thing (several times probably) before I hopefully understand it all.

Two things come to my mind:

1. This should probably be an appendix instead of being part of the main tutorial, because it's about the theory behind quantities, not about how to use them when writing applications.
2. I think you should put your name at the beginning (for instance, after the title), because this brilliant and you should get due credit for it.

What do you think?

**Comment by kiana-S** [comment]
@stefan-hoeck

>     1. This should probably be an appendix instead of being part of the main tutorial, because it's about the theory behind quantities, not about how to use them when writing applications.

I actually put it in the appendix section at first, but like I mentioned before I wasn't sure if it would fit there (that seems to mostly be for resources outside of learning the language itself).

>     2. I think you should put your name at the beginning (for instance, after the title), because this is brilliant and you should get due credit for it.

I wasn't really sure how to credit myself. Maybe a note in the introduction? I'll probably add something like that when I make the other changes.

**Comment by stefan-hoeck** [comment]
Sorry for the silence. I'm planning to truly work through this in order to try and understand everyghing. Might have some questions then.

**Comment by kiana-S** [comment]
@stefan-hoeck Any updates/questions on this chapter? It's been quite a while

**Comment by stefan-hoeck** [comment]
> @stefan-hoeck Any updates/questions on this chapter? It's been quite a while

Yes, I'm sorry. Working through it now. Thanks for your patience.

File: src/Appendices/QTT.md [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+ignore : a -> Int
+ignore x = 42
+```
+
+The input `x` is used zero times, which would normally mean its
+quantity would have to be `0`; however, since `0 ≤ ω`, we can use
+subusaging to increase the quantity to `ω`.
+
+This also explains the mysterious fact we pointed out earlier, that `0
+≰ 1` in our quantity ordering. If it were true that `0 ≤ 1`, then we
```
Comment (stefan-hoeck) [inline src/Appendices/QTT.md:272]: This should go to the previous line: `0 ≰ 1`, otherwise, the backticks are not properly closed.

**Comment by stefan-hoeck** [comment]
So, I finally managed to properly work through the whole text. The details about how quantities are calculated (especially in function application) took some time. In the pattern matching example, I *think* I know where each quantity comes from, but I'm not yet sure I got it right. That's the only section where I think it might be helpful to walk readers through the whole calculation one more time.

Other than that, I think this is a great write and a very welcome addition to the tutorial.

File: src/Appendices/QTT.md (hunk 1 above)
Comment (kiana-S) [reply src/Appendices/QTT.md:272]: Ah, thanks for catching that. I usually use soft line wrapping when editing text files and then apply an algorithmic line breaking system afterwards, which can sometimes mess up things.

**Comment by kiana-S** [comment]
Thanks for your feedback!

> In the pattern matching example, I think I know where each quantity comes from, but I'm not yet sure I got it right. That's the only section where I think it might be helpful to walk readers through the whole calculation one more time.

That makes sense. The section on pattern matching is the only part of this where I was deliberately light on details, for a few different reasons:

1. As mentioned in my note at the end, the explanation I gave was a heuristic that don't necessarily apply to all circumstances (it only applies to "simple" pattern matching, for a certain definition of that term).
2. Any more complete discussion of pattern matching in Idris would inevitably stumble on to the topic of pattern matching on the equality type, a subject that I intentionally avoided because I don't have any answers!

   (Somehow in the course of writing this chapter I couldn't find a single publication describing QTT that includes propositional equality. I think the Idris 2 compiler itself might genuinely be the first time that anyone has ever tried to do that.)

Because of those reasons, pattern matching was something I didn't want to linger on too much, and so the section suffered a bit because of that. I'll definitely go back and clean it up a bit.

**Comment by kiana-S** [comment]
Ah, I accidentally deleted the fork when I was moving my repos to a different remote. I'll see if I can fix this.

**Comment by kiana-S** [comment]
Unfortunately, it seems like I'll have to create a new PR...

**Pull request closed without merging**